crates/pathfinder/src/rpc/v02/types/class.rs

```diff
@@ -0,0 +1,178 @@
+            // ABIs are set by users and not verified by starknet, therefore ABIs
+            // can fail to parse (and just be nonsense). Discard these ABIs.
+            serde_json::from_value::<Vec<ContractAbiEntry>>(json).ok()
+        });
+
+        let program = json_obj
+            .get_mut("program")
+            .context("program property is missing")?
+            .to_string();
+
+        // Program is expected to be a gzip-compressed then base64 encoded representation of the JSON.
+        let mut gzip_encoder =
+            flate2::write::GzEncoder::new(Vec::new(), flate2::Compression::fast());
+        serde_json::to_writer(&mut gzip_encoder, &program).context("Compressing program")?;
+        let compressed_program = gzip_encoder
+            .finish()
+            .context("Finalizing program compression")?;
+        let encoded_program = base64::encode(compressed_program);
+        let program = encoded_program;
+
+        Ok(ContractClass {
+            program,
+            entry_points_by_type: entry,
+            abi,
+        })
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct ContractClass {
+    program: String,
+    entry_points_by_type: ContractEntryPoints,
+    abi: Option<Vec<ContractAbiEntry>>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+#[serde(deny_unknown_fields)]
+struct ContractEntryPoints {
+    constructor: Vec<ContractEntryPoint>,
+    external: Vec<ContractEntryPoint>,
+    l1_handler: Vec<ContractEntryPoint>,
+}
+
+#[serde_with::serde_as]
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct ContractEntryPoint {
+    #[serde_as(as = "U64AsHexStr")]
+    offset: u64,
+    selector: StarkHash,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+#[serde(deny_unknown_fields)]
+enum ContractAbiEntry {
+    Function(FunctionAbiEntry),
+    Event(EventAbiEntry),
+    Struct(StructAbiEntry),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "lowercase")]
+#[serde(deny_unknown_fields)]
+enum StructAbiType {
+    Struct,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "lowercase")]
+#[serde(deny_unknown_fields)]
+enum EventAbiType {
+    Event,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "snake_case")]
+#[serde(deny_unknown_fields)]
+enum FunctionAbiType {
+    Function,
+    L1Handler,
+    // This is missing from the v0.2 RPC specification and will be added in the
+    // next version. We add it as a deviation from the current spec, since it is
+    // effectively a bug in the v0.2 specification.
+    Constructor,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct StructAbiEntry {
+    r#type: StructAbiType,
+    name: String,
+    size: u64,
+    members: Vec<StructMember>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct StructMember {
+    #[serde(flatten)]
```

Comment:
Unfortunately this is not supported by serde. Quoting from https://serde.rs/attr-flatten.html:

> Note: flatten is not supported in combination with structs that use [deny_unknown_fields](https://serde.rs/container-attrs.html#deny_unknown_fields). Neither the outer nor inner flattened struct should use that attribute.

Reply:
Thanks - I think I fixed this in my latest commit.